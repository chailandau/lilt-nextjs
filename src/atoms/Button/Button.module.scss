.button {
    display: flex;
    justify-content: center;
    align-items: center;
    border-radius: rem(20);
    padding: rem(12) rem(20);
    @include font(14, 16);
    @include font-rubik-bold;
    color: color(neutral, black);
    text-transform: uppercase;
    flex-grow: 1;
    transition: flex-grow 0.3s ease-in-out;
    @include tablet {
        flex-grow: 0;
    }
    &.blue {
        @include bgSlideHover(color(neon, blue), color(neutral, white));
    }
    &.green {
        @include bgSlideHover(color(neon, green), color(neutral, white));
    }
    &.white {
        @include bgSlideHover(color(neutral, white), color(neon, blue));
    }
}
