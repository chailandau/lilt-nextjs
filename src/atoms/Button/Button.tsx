import classNames from 'classnames';
import { FC } from 'react';

import styles from './Button.module.scss';

import parseUrl from '@/utils/parseUrl';

export const buttonColors = ['blue', 'green', 'white'] as const;

interface ButtonProps {
    /** Button color */
    color?: typeof buttonColors[number];
    /** Button content */
    children: string;
    /** `Button` renders as an `a` tag if supplied */
    link?: string;
    /** `onClick` event */
    onClick?: () => void;
}

const Button: FC<ButtonProps> = ({
    color = 'blue',
    link,
    onClick,
    children
}) => {

    const classes = classNames(styles['button'], styles[color]);

    const isLink = link && link.trim() !== '';

    const {
        rel = undefined,
        target = undefined,
        href = undefined
    } = (isLink && parseUrl(link)) || {};

    const ButtonTag = isLink ? 'a' : 'button';

    return (
        <ButtonTag
            href={isLink ? href : undefined}
            rel={rel}
            target={target}
            className={classes}
            onClick={onClick}
        >
            {children}
        </ButtonTag>
    );
};

export default Button;
