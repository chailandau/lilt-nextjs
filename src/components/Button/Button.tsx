import { FC } from 'react';

import { Button as ButtonType } from '@/api/graphqlTypes';
import ButtonMolecule, {
    ButtonMoleculeVariant
} from '@/molecules/ButtonMolecule/ButtonMolecule';

export interface ButtonProps {
    /** Optional classname */
    className?: string;
    /** Button object */
    buttonData: ButtonType | null;
}

const Button: FC<ButtonProps> = ({ buttonData, className }) => {
    const { linkType, color, label, externalLink, internalLink } =
        buttonData || {};

    const isExternal = linkType === 'external';

    const getLink = () => {
        if (isExternal && externalLink) {
            return `${externalLink}`;
        } else if (internalLink?.slug) {
            return `${process.env.NEXT_PUBLIC_BASE_URL}/${internalLink.slug}`;
        } else {
            return '';
        }
    };

    const buttonProps = {
        variant: color as ButtonMoleculeVariant,
        className
    };

    return (
        label && (
            <ButtonMolecule link={getLink()} {...buttonProps}>
                {label}
            </ButtonMolecule>
        )
    );
};

export default Button;
