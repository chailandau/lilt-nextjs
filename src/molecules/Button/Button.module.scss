.button {
    display: flex;
    justify-content: center;
    align-items: center;
    border-radius: rem(20);
    padding: rem(12) rem(20);
    @include font(14, 16);
    @include font-rubik-bold;
    color: color(neutral, black);
    text-transform: uppercase;
    flex-grow: 1;
    transition: flex-grow 0.3s ease-in-out;
    @include mobile-lg {
        flex-grow: 0;
    }
    &.blue {
        background-color: color(neon, blue);
    }
    &.green {
        background-color: color(neon, green);
    }
    &.white {
        background-color: color(neutral, white);
    }
}
