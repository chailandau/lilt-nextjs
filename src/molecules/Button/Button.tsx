import classNames from 'classnames';
import { FC } from 'react';

import styles from './Button.module.scss';

import Link from '@/atoms/Link/Link';

export const buttonVariants = ['blue', 'green', 'white', 'link'] as const;

export interface ButtonProps {
    /** Button variant */
    variant?: (typeof buttonVariants)[number];
    /** Button content */
    children: string;
    /** `Button` renders as an `a` tag if supplied */
    link?: string;
    /** `onClick` event */
    onClick?: () => void;
    /** `onMouseEnter` event */
    onMouseEnter?: () => void;
    /** `onMouseLeave` event */
    onMouseLeave?: () => void;
    /** CSS `className` */
    className?: string;
}

const Button: FC<ButtonProps> = ({
    variant = 'blue',
    link,
    onClick,
    onMouseEnter,
    onMouseLeave,
    children,
    className
}) => {
    const classList = classNames(
        styles['button'],
        styles[variant],
        className && className
    );

    const isLink = link && link.trim() !== '';

    return isLink ? (
        <Link href={link} className={classList} underline={false}>
            {children}
        </Link>
    ) : (
        <button
            className={classList}
            onClick={onClick}
            onMouseEnter={onMouseEnter}
            onMouseLeave={onMouseLeave}
        >
            {children}
        </button>
    );
};

export default Button;
