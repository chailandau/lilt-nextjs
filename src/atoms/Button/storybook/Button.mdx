import { Canvas, Meta, ArgTypes } from "@storybook/blocks";

import * as ButtonStories from "./Button.stories";

<Meta of={ButtonStories} />

# Button Atom

<Canvas of={ButtonStories.Default} />

<br />

### `children`

Text is supplied as a child to the `Button` atom.

### `color`

The color of the `Button` is controlled by the `color` prop.

### `link`

If the link prop is not empty, the `Button` will render as an `a` tag. Otherwise it will default to a `button` tag.

### `onClick`

If the `onClick` prop is supplied, it will be called when the `Button` is clicked.

## Props

<ArgTypes />
