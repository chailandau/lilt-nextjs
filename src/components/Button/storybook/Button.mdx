import { Canvas, Meta, ArgTypes } from "@storybook/blocks";

import * as ButtonStories from "./Button.stories";

<Meta of={ButtonStories} />

# Button Atom

<Canvas of={ButtonStories.Default} />

<br />

### children

Text is supplied as a child to the `Button` molecule.

<hr />

### color

Controls the color of the `Button` molecule.

<hr />

### link

If the link prop is not empty, the `Button` will render as an `a` tag. Otherwise it will default to a `button` tag.

<hr />

### onClick

If the `onClick` prop is supplied, it will be called when the `Button` is clicked.

<br />
<br />
<br />

## Props

<ArgTypes />
