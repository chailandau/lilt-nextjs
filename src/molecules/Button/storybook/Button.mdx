import { Canvas, Meta, ArgTypes } from '@storybook/blocks';

import * as ButtonStories from './Button.stories';

<Meta of={ButtonStories} />

# Button Molecule

<Canvas of={ButtonStories.Default} />

<br />

## children

Text is supplied as a child to the `Button` molecule.

## color

Controls the color of the `Button` molecule.

## link

If the link prop is not empty, the `Button` will render as an `a` tag. Otherwise, it will default to a `button` tag.

## onClick

If there is no link and the `onClick` prop is supplied, it will be called when the `Button` is clicked.

<br />
<br />
<br />

## Props Table

<ArgTypes />
