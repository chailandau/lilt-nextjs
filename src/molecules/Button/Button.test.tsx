import { composeStory } from '@storybook/react';
import { render, screen } from '@testing-library/react';

import { buttonVariants } from './Button';
import Meta, { Default } from './storybook/Button.stories';

import {
    testAppliesClass,
    testAxeViolations,
    testKeyDown,
    testMatchesSnapshot,
    testPropOptions,
    testRenderText
} from '@/utils/testHelpers';

const Button = composeStory(Default, Meta);

describe('Button', () => {
    const buttonComponent = {
        component: <Button />
    };
    testRenderText({
        ...buttonComponent,
        role: 'button',
        text: Button.args.children as string
    });
    it('has default color', () => {
        render(<Button variant={undefined} />);
        expect(screen.getByRole('button')).toHaveClass('blue');
    });
    testPropOptions({
        component: <Button />,
        propName: 'variant',
        propOptions: buttonVariants,
        htmlTag: 'button'
    });
    it("renders 'link' correctly", () => {
        render(<Button link='https://google.com' />);
        expect(screen.getByRole('link').getAttribute('href')).toBe(
            'https://google.com'
        );
    });
    it("renders 'button' correctly", () => {
        render(<Button />);
        expect(screen.getByRole('button')).toBeDefined();
    });
    testKeyDown({ ...buttonComponent, role: 'button' });
    testAppliesClass({ ...buttonComponent, htmlTag: 'button' });
    testAxeViolations(buttonComponent);
    testMatchesSnapshot(buttonComponent);
});
