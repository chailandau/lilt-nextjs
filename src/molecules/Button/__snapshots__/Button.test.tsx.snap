// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button has 'blue' class 1`] = `
<div
  className="flex"
>
  <button
    className="button blue"
    onKeyDown={[Function]}
  >
    Default button
  </button>
</div>
`;

exports[`Button has 'green' class 1`] = `
<div
  className="flex"
>
  <button
    className="button blue"
    onKeyDown={[Function]}
  >
    Default button
  </button>
</div>
`;

exports[`Button has 'link' class 1`] = `
<div
  className="flex"
>
  <button
    className="button blue"
    onKeyDown={[Function]}
  >
    Default button
  </button>
</div>
`;

exports[`Button has 'white' class 1`] = `
<div
  className="flex"
>
  <button
    className="button blue"
    onKeyDown={[Function]}
  >
    Default button
  </button>
</div>
`;

exports[`Button matches snapshot 1`] = `
<div
  className="flex"
>
  <button
    className="button blue"
    onKeyDown={[Function]}
  >
    Default button
  </button>
</div>
`;
